components/sync15/src/engine/bridged_engine.rs

```diff
@@ -87,8 +90,118 @@ pub trait BridgedEngine: Send + Sync {
+// from the explicit timestamp handling and moving closer to the `get_collection_request`
+// model.
+pub trait BridgedEngineAdaptor: Send + Sync {
+    // These are the main mismatches between the 2 engines
+    fn last_sync(&self) -> Result<i64>;
+    fn set_last_sync(&self, last_sync_millis: i64) -> Result<()>;
+    fn sync_started(&self) -> Result<()> {
+        Ok(())
+    }
+
+    fn engine(&self) -> &dyn SyncEngine;
+}
+
+impl<A: BridgedEngineAdaptor> BridgedEngine for A {
+    fn last_sync(&self) -> Result<i64> {
+        self.last_sync()
+    }
+
+    fn set_last_sync(&self, last_sync_millis: i64) -> Result<()> {
+        self.set_last_sync(last_sync_millis)
+    }
+
+    fn sync_id(&self) -> Result<Option<String>> {
+        Ok(match self.engine().get_sync_assoc()? {
+            EngineSyncAssociation::Disconnected => None,
+            EngineSyncAssociation::Connected(c) => Some(c.coll.into()),
+        })
+    }
+
+    fn reset_sync_id(&self) -> Result<String> {
+        let global = Guid::empty();
+        let coll = Guid::random();
+        self.engine()
+            .reset(&EngineSyncAssociation::Connected(CollSyncIds {
+                global,
+                coll: coll.clone(),
+            }))?;
+        Ok(coll.to_string())
+    }
+
+    fn ensure_current_sync_id(&self, sync_id: &str) -> Result<String> {
+        let engine = self.engine();
+        let assoc = engine.get_sync_assoc()?;
+        if matches!(assoc, EngineSyncAssociation::Connected(c) if c.coll == sync_id) {
+            log::debug!("ensure_current_sync_id is current");
+        } else {
+            let new_coll_ids = CollSyncIds {
+                global: Guid::empty(),
+                coll: sync_id.into(),
+            };
+            engine.reset(&EngineSyncAssociation::Connected(new_coll_ids))?;
+        }
+        Ok(sync_id.to_string())
+    }
+
+    fn prepare_for_sync(&self, client_data: &str) -> Result<()> {
+        // unwrap here is unfortunate, but can hopefully go away if we can
+        // start using the ClientData type instead of the string.
+        self.engine()
+            .prepare_for_sync(&|| serde_json::from_str::<crate::ClientData>(client_data).unwrap())
+    }
+
+    fn sync_started(&self) -> Result<()> {
+        A::sync_started(self)
+    }
+
+    fn store_incoming(&self, incoming_records: Vec<IncomingBso>) -> Result<()> {
+        let engine = self.engine();
+        let mut telem = telemetry::Engine::new(engine.collection_name());
+        engine.stage_incoming(incoming_records, &mut telem)
+    }
+
+    fn apply(&self) -> Result<ApplyResults> {
+        let engine = self.engine();
+        let mut telem = telemetry::Engine::new(engine.collection_name());
+        // we tell it to apply with an invalid timestamp then update it.
+        // this isn't atomic and we probably shouldn't do it, but to get a feel...
+        let records = engine.apply(ServerTimestamp::from_millis(0), &mut telem)?;
```

Comment:
Can you remind me why this isn't atomic, please? I paged this out of my brain 😅

It is wrong, but the only engines we have to worry about are tabs (and Extension Storage, maybe? 🤔), right? And it looks like its `apply` doesn't use the `timestamp`, so I think it's okay to leave in until we migrate!